Sprint_4

w33ka opened the pull request:
Задания выполнены

w33ka committed:
Готовое задание по проекту 4ого спринта

w33ka committed:
Исправление README и теста test_get_books_for_children

kadirovgm commented on README.md:
```diff
@@ -1 +1,34 @@
\ No newline at end of file
+# qa_python
+# СПИСОК ПРОВЕРОК
+
```
отлично!

kadirovgm commented on conftest.py:
```diff
@@ -0,0 +1,11 @@
+    return BooksCollector()
+
+
+def pytest_make_parametrize_id(config, val):
```
нужно исправить: не совсе понял для чего это?

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+#проверка на добавление книги без жанра и получение текщего словаря books_genre
+def test_add_new_book_without_genre(book_collector):
+    book_collector.add_new_book('Гарри Поттер')
+    book_collector.add_new_book('Властелин Колец')
```
нужно исправить: для юнит теста достаточно одного положительного значения с данным из эквивалентного класса, поэтому тут и добавления одной книги достаточно

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+    book_collector.add_new_book('Властелин Колец')
+
+    result = book_collector.get_books_genre()
+    assert len(result) != 0
```
нужно исправить: это некорректная проверка, так как значение не равно нулю, не означает что книг добавилось одно или две, их могло добавиться и 10, если вдруг из-за чего то они задублировались

поэтому стоит явно проверить, что длинa словаря стала равна 1, или же сделать проверку вида: `assert book in book_collector.get_books_genre()`

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+
+
+#проверка на добавление книги с названием больше 40 символов
+def test_add_new_book_without_genre_more_than_40symbols(book_collector):
```
можно лучше: тут можно добавить параметризацию, так как помимо длины больше 40, еще есть проверка на то, что книга с длиной 0 также не может быть добавлена. Поэтому парамтеры из разных классов эквивалентности 0 и 41 могут быть переданы в тест, и на выходе будем ожидать что в словарь книги не добавились

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+    book_collector.add_new_book('БольшаяКнигаНастолькоЧтоНеУмещаетсяСюдаОй')
+
+    result = book_collector.get_books_genre()
+    assert 'БольшаяКнигаНастолькоЧтоНеУмещаетсяСюдаОй' not in result
```
супер!

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+def test_set_book_genre(book_collector):
+    book_collector.add_new_book('Гарри Поттер')
+    book_collector.set_book_genre('Гарри Поттер', 'Фантастика')
+    assert book_collector.get_book_genre('Гарри Поттер') == 'Фантастика'
```
отлично, хороший тест!

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+
+    result = book_collector.get_books_genre()
+    assert 'БольшаяКнигаНастолькоЧтоНеУмещаетсяСюдаОй' not in result
+
```
можно лучше: также можешь добавить параметризированный тест  на проверку граничных значение: 1 и 40 - книги с таким количеством символов должны добавиться в словарь

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+    assert 'Красная шапочка' in children_book and 'Колобок' not in children_book
+
+# проверка на добавление книги в избранное из словаря books_genre  и проверка на вывод списка избранного
+@pytest.mark.parametrize('book_name', ['Винни Пух', 'Буратино'])
```
нужно исправить: думаю тут параметризация неуместна

параметризация - это когда на вход теста передают тестовые данные из разных классов эквивалентности, при этом зная, что тест вернет одно и то же значение

нам достаточно на одном из положительных значений проверить, что добавление в избранные работает

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+    book_collector.add_new_book('Винни Пух')
+    book_collector.add_book_in_favorites('Винни Пух')
+
+    favorites = book_collector.get_list_of_favorites_books()
```
нужно исправить: чем этот тест отличается от test_add_book_in_favorites ?

также тест должны быть атомарны один тест=одна проверка (один ассерт), так что эти строки бы я вообще убрал и оставил только проверку на то что книги не дублируются в списке favorites

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,90 @@
+#проверка на удаление книги из избранного
+def test_delete_book_from_favorites(book_collector):
+    book_collector.add_new_book('Белоснежка и семь гномов')
+    book_collector.add_new_book('Маленький принц')
```
нужно исправить: для чего тут добавление книги `Маленький принц` ? кажется это никак не влияет на тест

w33ka committed:
Исправление conftest.py и tests.py согласно замечаниям

w33ka committed:
Исправление tests.py согласно замечаниям:
1.Добавлена параметризация в тест test_add_new_book_without_genre_more_than_40symbols_or_null
2.Добавлен тест test_add_new_book_with_one_or_forty_symbols с параметризацией

Исправление README.md в связи с добавлением нового теста

kadirovgm commented on tests.py:
```diff
@@ -1,24 +1,96 @@
+    book_collector.add_new_book(book_name)
+
+    result = book_collector.get_books_genre()
+    assert book_name not in result
```
отлично!
